### ISSUE TITLE
`Result::unwrap()` on an `Err` value: Conversion("cannot interpret F64(Some(0.0)) as an f32 value")

### ISSUE THREAD

**Issue by Pyzyryab** [prisma/tiberius#249]
Hello. I am using tiberius on some database ORM, and I am unable to play with float types. 

Whenever I define a classical float on SqlServer, tiberius always broke the result on the `row.get()` method.

Is this a know bug? Whatching the crate `tiberius-derive`, they specifically talk about this one.
```
Known issues:

There is a tiberius bug(?) that breaks Option fields at the moment. Tiberius appears to always parse null Float(53) (double) columns as Nullable(f32) independently of the column type.
```

So, my Rust field is an Option<f64>. If I try to change it to an Option<f32>...
```
Result::unwrap()` on an `Err` value: Conversion("cannot interpret F32(None) as an f64 value")
```

Is this a known bug? Exists some workaround over it? How can I fix my code? I am using it on an enterprise level application, releasing the code in production, and we're getting delayed because we can't work with our float columns.

Thanks in advice for your time.

**Comment by Pyzyryab** [comment]
I downloaded the `tiberius` source-code. Error is on the `fron_sql()!` macro for floats.

I manually implemented `FromSql<'a>` for f32 and f64.
Having a field determined as an `Option<i64>`, I get these results:
```
Data as F64: Some(11.3)
Data as F64: Some(11.3)
Data as F64: Some(11.5)
Data as F64: Some(15.5)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(10.0)
Data as F64: Some(0.0)
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as F64: Some(0.0)
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as F64: Some(0.0)
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as F64: Some(0.0)
Data as f32: None
Data as F64: Some(0.0)
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as F64: Some(0.0)
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as F64: Some(0.0)
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as F64: Some(0.0)
Data as F64: Some(0.0)
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
Data as f32: None
```
On an Option<f64> Rust variable, tiberius is retrieving from an `SqlServer` column defined as `float - null`, values when
present as `f64`, and null values as optional of `f32`.

I literally just matched against bot `ColumnData` of `F32` and `F64` variants. Error should be more deep that the `FromSql` impl from floats. I have not enought time to investigate it, but I hope that this leads to the tiberius dev team to find such bug. 

Meanwhile, I will open a PR for solving the issues. Code it's provisional, as it shouldn't be a full correction, but at least
tiberius now is able to play with Option<f32> and Option<f64> values.

I would like to work with you when I have some time. I also would like that the dev team take my PR and performn the full correction over it, so I can contribute to the `tiberius` ecosystem, and be aware of the solution implemented by you.

**Comment by Pyzyryab** [comment]
So, obv, I have no permissions to push to this repository. I have the initial solution on a new branch, and I would like to contribute. Is this possible?

**Comment by pimeys** [comment]
Hey, we have a holiday today so probably not answering super fast. Send a PR if you think you found a bug!

At least [here](https://github.com/prisma/tiberius/blob/main/src/tds/codec/column_data/float.rs#L9-L19) we do detect the float size even for nulls. And [here](https://github.com/prisma/tiberius/blob/main/src/from_sql.rs#L65-L66) we provide conversion functions for both: `f32` and `f64`, including nulls.

**Comment by Pyzyryab** [comment]
Hello @pimeys thanks for your response even on holidays. I am forking the project now to make a pull request. Solution provided is not beatiful, but stops the breakage when a float value is required. I will make the PR and the dev-team can take a look at it. 

Here is exactly the problem. Float null values are getting parsed as `ColumnData::F32`, and any non null value it's being retrieved as `ColumnData::F64`. That's such an extrange behaviour, and I guess that is a deep bug (I mean, when bytes are read on the query response).

I am making the PR now

**Comment by pimeys** [comment]
If the float is nullable, it should come back as variable length column. The column type info tells us the size and we act accordingly.

I remember similar issues being fixed in this pull request: https://github.com/prisma/tiberius/pull/209

Are you running the latest version?

**Comment by Pyzyryab** [comment]
Yes, I upgraded tiberius to `0.11.2`. I guess that is easy to check the bug with some `SqlServer` instance, and some column defined as `float`, obv, as nullable. I can provide logs or whatever if you're on holidays, and make the debug work for you

**Comment by pimeys** [comment]
Just a test would be great!

**Comment by Pyzyryab** [comment]
I am working with (Canyon-SQL)(https://github.com/zerodaycode/Canyon-SQL), that acts as an ORM. Firstly was integrated with `tokio_postgres`, now with `tiberius`.

A simple struct to use with `Canyon-SQL`, with a datasource pointing to an `SqlServer` instance, which will trigger the `tiberius` connector. 

```
#[derive(Debug, Clone, CanyonCrud, CanyonMapper)]
#[canyon_macros::canyon_entity(table_name = "Ticket", schema = "tic")]
pub struct Ticket {
    #[primary_key]
    pub TicketCode: i64,
    pub DedicatedHours: Option<f64>,
}
```

Canyon uses as dependency: `tiberius = { version = "0.11.2", features = ["tds73", "chrono"] }`

Now, a query is generated, gets the results back as a `Vec<tiberius::Row>`, and the `CanyonMapper` macro just inits the struct fields and uses:

```
quote! {  
    #ident: row.get::<#ty, &str>(#ident_name)
        .expect(format!("Failed to retrieve the `{}` field", #ident_name).as_ref())
 }
```
there's a lot more of usecases for diff types, just trying to set up a decent explanation.

After that, when a field contains an `Option<f64>`:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Conversion("cannot interpret F32(None) as an f64 value")'
```

and if I change it to an `Option<f32>`:
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Conversion("cannot interpret F64(Some(11.3)) as an f32 value")'
```

After manually implementing `FromSql<'a> for both `f32` and `f64, and as commented above, *tiberius* is matching the 
`ColumnData::F32` on None values, and `ColumnData::F64` on any non null value. 

```
impl<'a> FromSql<'a> for f64 {
    fn from_sql(value: &'a ColumnData<'static>) -> crate::Result<Option<Self>> {
        match value {
            ColumnData::F32(data) => {  // Here are always the None values
                Ok(data.map( |v| v as f64))
            },
            ColumnData::F64(data) => {  // Here are always the non null-values
                Ok(*data)
            },
            v => {
                Err(crate::Error::Conversion(
                    format!("cannot interpret {:?} as an f64 value", v).into(),
                ))
            }
        }
    }
}
```

I guess that (this)[https://github.com/prisma/tiberius/blob/main/src/tds/codec/column_data/float.rs#L9-L19] fix is not solving the issue.

**Comment by pimeys** [comment]
What is the exact table definition in the database? Nullable reals or nullable floats?

**Comment by Pyzyryab** [comment]
nullable float

**Comment by pimeys** [comment]
I wrote a few tests, but I can't really reproduce this:

```rust
#[test_on_runtimes]
async fn nullable_f64<S>(mut conn: tiberius::Client<S>) -> Result<()>
where
    S: AsyncRead + AsyncWrite + Unpin + Send,
{
    let table = random_table().await;

    conn.execute(format!("CREATE TABLE ##{table} (m1 float NULL)"), &[])
        .await?;

    conn.execute(&format!("INSERT INTO ##{table} (m1) VALUES (null)"), &[])
        .await?;

    let row = conn
        .query(&format!("SELECT m1 FROM ##{table}"), &[])
        .await?
        .into_row()
        .await?
        .unwrap();

    let col: Option<f64> = row.get(0);
    assert_eq!(None, col);

    Ok(())
}

#[test_on_runtimes]
async fn nullable_f32<S>(mut conn: tiberius::Client<S>) -> Result<()>
where
    S: AsyncRead + AsyncWrite + Unpin + Send,
{
    let table = random_table().await;

    conn.execute(format!("CREATE TABLE ##{table} (m1 real NULL)"), &[])
        .await?;

    conn.execute(&format!("INSERT INTO ##{table} (m1) VALUES (null)"), &[])
        .await?;

    let row = conn
        .query(&format!("SELECT m1 FROM ##{table}"), &[])
        .await?
        .into_row()
        .await?
        .unwrap();

    let col: Option<f32> = row.get(0);
    assert_eq!(None, col);

    Ok(())
}
```

Both of these are green with the latest `main` branch.

**Comment by Pyzyryab** [comment]
can you point me where do you wrote those tests? So I can download them to my environment and check them out?

**Comment by pimeys** [comment]
https://github.com/prisma/tiberius/pull/251

Have one of the SQL Servers running defined in the docker compose setup, run `cargo test nullable_f`.

**Comment by pimeys** [comment]
Maybe try to modify the tests if they're doing something different than you do. The CI runs all the tests, so you can take a look into `.github` directory and see how we initialize the test env.

**Comment by Pyzyryab** [comment]
Fine. I'll give it a look between today and tomorrow. Sorry for the delay time.

**Comment by Pyzyryab** [comment]
Hello, @pimeys may I ask for help with this error?

```
[2022-10-05T10:33:24Z ERROR tiberius::tds::stream::token] Windows logins are not supported in this version of SQL Server. code=40607
```

**Comment by pimeys** [comment]
The server you are using is not allowing windows logins. If you're using the docker image, it is not working with windows credentials.

For SQL Server authentication, and with the docker image defined in the `docker-compse.yml`, you should just source the `.envrc` file or set the env var:

```
export TIBERIUS_TEST_CONNECTION_STRING="server=tcp:localhost,1433;user=SA;password=<YourStrong@Passw0rd>;TrustServerCertificate=true"
```

You probably have something wrong in your connection string, e.g. this one uses the Windows integrated login:

```
server=tcp:127.0.0.1,1433;IntegratedSecurity=true;TrustServerCertificate=true
```

**Comment by Pyzyryab** [comment]
Well, test runned against our Azure SqlServer and passed.

But, for testing, I made a minimal environmet against the same database.

```
#[async_std::main]
async fn main() -> anyhow::Result<()> {
    let mut sqlserver_client = get_sqlserver_client().await;

    let mut select = Query::new("SELECT * FROM tic.Ticket WHERE CodTicket > @P1 AND CodTicket < @P2");
    select.bind(18);
    select.bind(25);

    let stream = select.query(&mut sqlserver_client).await?;
    let row = stream.into_results().await;

    for vector in row.into_iter() {
        for row in vector.into_iter() {
            for inner_row in row.into_iter() {
                println!("Column: {:?}", &inner_row.try_get::<f64, &str>("HorasDedicadas"));
            }
        }
    }

    Ok(())
}
```

Here are the results: 
```
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
```

Here's the first 30 entries for the database table with the `HorasDedicadas` values.

```
0
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
NULL
0
NULL
NULL
NULL
NULL
NULL
NULL
NULL
```

So, in the range of the binded parameters... there's nulls. No idea about what's going on.

**Comment by Pyzyryab** [comment]
So now, I decided to update with `NULL` the first 30 entries of the table (the bind parameters for the select query are now in range 0 - 31).

```
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Ok(None)
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Ok(None)
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
Column: Err(Conversion("cannot interpret F32(None) as an f64 value"))
```

**Comment by Pyzyryab** [comment]
Now, I updated all the same 30 values with some random float value (12.1).

Everything is fine: `Column: Ok(Some(12.1))`

We tested with multiple random float values on the update clause, an never brokes.

The issue is that, some `NULL` are being read as `F32(None)`, and some are being read as `F64(None)`. For an `Option<i64>`, those who are read as `F32(None)` obv fails. 

Same case is applied for a Rust `Option<f32>`, where the ``F32(None)` are valid, but some nulls are retrieved for some strange reason as `F64(None)`, which brokes the FromSql impl for float types.
